File: frontend/types/guardrail.ts
export type Decision = "ALLOW" | "BLOCK" | "ESCALATE";
export type TriggeredLayer =
  | "rule_based"
  | "grok_classifier"
  | "safety"
  | "router"
  | "output_guardrail"
  | "none";

export interface GuardrailResult {
  decision: Decision;
  category: string;
  reason: string;
  risk_score: number;
  confidence: number;
  triggered_layer: TriggeredLayer;
  layer_name: string;
  latency_ms: number;
  called: boolean;
}

export interface SafetyResult {
  category: "SAFE" | "SUSPICIOUS" | "HARMFUL" | "SENSITIVE" | "HUMAN_REVIEW";
  reason: string;
  risk_score: number;
  confidence: number;
  component: string;
  latency_ms: number;
  called: boolean;
}

export interface RouterResult {
  decision: Decision;
  category: string;
  reason: string;
  risk_score: number;
  confidence: number;
  triggered_layer: TriggeredLayer;
}

export interface PipelineTrace {
  request_id: string;
  input: string;
  rule_based: GuardrailResult;
  grok_classifier: GuardrailResult | null;
  safety: SafetyResult | null;
  router: RouterResult;
  assistant_output: string | null;
  output_guardrail: GuardrailResult | null;
  final_response: string;
  total_latency_ms: number;
  timestamp: string;
}

export interface AttackCase {
  id: string;
  input: string;
  expected_decision: Decision;
  category: string;
  expected_reason: string;
  label: string;
}

export interface AttackRunResult {
  case: AttackCase;
  actual_decision: Decision;
  actual_category: string;
  actual_reason: string;
  triggered_layer: string;
  passed: boolean;
  trace: PipelineTrace;
}

export interface EvaluationMetrics {
  total_requests: number;
  allow_count: number;
  block_count: number;
  escalate_count: number;
  attack_blocking_rate: number;
  false_refusal_rate: number;
  routing_accuracy: number;
  overall_accuracy: number;
  average_latency_ms: number;
}

export interface HealthResponse {
  status: string;
  grok_configured: boolean;
  guardrails_status: string;
}

File: backend/app/schemas/guardrail.py
from typing import Literal, Optional
from pydantic import BaseModel, Field

Decision = Literal["ALLOW", "BLOCK", "ESCALATE"]
TriggeredLayer = Literal[
    "rule_based", "grok_classifier", "safety", "router", "output_guardrail", "none"
]


class GuardrailResult(BaseModel):
    """Common structured output every guardrail layer produces."""
    decision: Decision
    category: str
    reason: str
    risk_score: float = Field(ge=0.0, le=1.0)
    confidence: float = Field(ge=0.0, le=1.0)
    triggered_layer: TriggeredLayer
    layer_name: str
    latency_ms: float = 0.0
    called: bool = True


class SafetyResult(BaseModel):
    category: Literal["SAFE", "SUSPICIOUS", "HARMFUL", "SENSITIVE", "HUMAN_REVIEW"]
    reason: str
    risk_score: float = Field(ge=0.0, le=1.0)
    confidence: float = Field(ge=0.0, le=1.0)
    component: str
    latency_ms: float = 0.0
    called: bool = True


class RouterResult(BaseModel):
    decision: Decision
    category: str
    reason: str
    risk_score: float
    confidence: float
    triggered_layer: TriggeredLayer


class PipelineTrace(BaseModel):
    """Full record of a single request moving through every layer."""
    request_id: str
    input: str
    rule_based: GuardrailResult
    grok_classifier: Optional[GuardrailResult] = None
    safety: Optional[SafetyResult] = None
    router: RouterResult
    assistant_output: Optional[str] = None
    output_guardrail: Optional[GuardrailResult] = None
    final_response: str
    total_latency_ms: float = 0.0
    timestamp: str
